[LoopVectorize] Teach LoopVectorizationLegality about more early exits

david-arm opened the pull request:
This patch is split off from PR #88385 and concerns only the code related to the legality of vectorising early exit loops. It is the first step in adding support for vectorisation of a simple class of loops that typically involves searching for something, i.e.

```
  for (int i = 0; i < n; i++) {
    if (p[i] == val)
      return i;
  }
  return n;
```

or

```
  for (int i = 0; i < n; i++) {
    if (p1[i] != p2[i])
      return i;
  }
  return n;
```

In this initial commit LoopVectorizationLegality will only consider early exit loops legal for vectorising if they follow these criteria:

1. There are no stores in the loop.
2. The loop must have only one early exit like those shown in the above example. I have referred to such exits as speculative early exits, to distinguish from existing support for early exits where the exit-not-taken count is known exactly at compile time.
3. The early exit block dominates the latch block.
4. The latch block must have an exact exit count.
5. There are no loads after the early exit block.
6. The loop must not contain reductions or recurrences. I don't see anything fundamental blocking vectorisation of such loops, but I just haven't done the work to support them yet.
7. We must be able to prove at compile-time that loops will not contain faulting loads.

Tests have been added here:

  Transforms/LoopVectorize/AArch64/simple_early_exit.ll

david-arm committed:
[LoopVectorize] Teach LoopVectorizationLegality about more early exits

This patch is split off from PR #88385 and concerns only the code
related to the legality of vectorising early exit loops. It is the
first step in adding support for vectorisation of a simple class of
loops that typically involves searching for something, i.e.

  for (int i = 0; i < n; i++) {
    if (p[i] == val)
      return i;
  }
  return n;

or

  for (int i = 0; i < n; i++) {
    if (p1[i] != p2[i])
      return i;
  }
  return n;

In this initial commit LoopVectorizationLegality will only consider
early exit loops legal for vectorising if they follow these criteria:

1. There are no stores in the loop.
2. The loop must have only one early exit like those shown in the above
example. I have referred to such exits as speculative early exits, to
distinguish from existing support for early exits where the
exit-not-taken count is known exactly at compile time.
3. The early exit block dominates the latch block.
4. The latch block must have an exact exit count.
5. There are no loads after the early exit block.
6. The loop must not contain reductions or recurrences. I don't see
anything fundamental blocking vectorisation of such loops, but I just
haven't done the work to support them yet.
7. We must be able to prove at compile-time that loops will not contain
faulting loads.

Tests have been added here:

  Transforms/LoopVectorize/AArch64/simple_early_exit.ll

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1442,6 +1486,95 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+          NumExitBlocks++;
+        }
+      }
+      if (NumExitBlocks > 1) {
```
Switch support has only be added recently, so would probably be good to add a test if possible?

david-arm replied:
I think I already have with `@early_exit_has_multiple_outside_successors`, unless you were thinking of a different test?

paulwalker-arm commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -2445,8 +2445,11 @@ bool LoopAccessInfo::analyzeLoop(AAResults *AA, const LoopInfo *LI,
-      if (Call && getVectorIntrinsicIDForCall(Call, TLI))
-        continue;
+      if (Call) {
+        NumCalls++;
```
Based on the use, is this too broad? For example, would it prevent loops that call sqrt intrinsics from being vectorisable?  What about keeping a count of non-store instructions that can modify memory? This can be added to the code below.

david-arm replied:
Hmm, I was probably being overly cautious here, but I'm worried about more than just stores. The function call could have side-effects, such as changing system state or generating exceptions. I might be able to relax this to permit calls to intrinsics that have no side-effects, and do not write to memory. I'll see how easy this is to do - it would require adding new tests.

david-arm replied:
This now makes me wonder about operations such as integer divides that may generate exceptions and whether or not such operations should be considered legal in an early exit block.

fhahn replied:
I don't think we can allow any instruction that may trigger UB or a trap in the header if it would be executed unconditionally in the vectorized loop, but this should probably be checked outside of LAA, as it is irrelevant for memory dependence analysis.

david-arm replied:
Agreed. It might make more sense to analyse all instructions in the early exiting block by checking for I.mayHaveSideEffects(), I.mayThrow(), I.mayReadOrWriteMemory(), etc.

david-arm replied:
I've reverted this code in favour of calling `isSafeToSpeculativelyExecute` from LoopVectorizationLegality. This covers more cases such as divides, while at the same time permitting calls to intrinsics such as sqrt.

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1070,6 +1070,50 @@ bool LoopVectorizationLegality::canVectorizeMemory() {
+    }
+
+    // The vectoriser cannot handle loads that occur after the early exit block.
+    BasicBlock *LatchBB = TheLoop->getLoopLatch();
+    for (Instruction &I : *LatchBB) {
```
~This seems like a naive question but is the latch block the only one that needs to be check? Are there no other blocks between the existing an latch blocks?~

I answered my own question but leaving comment here because it is referenced below.

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1442,6 +1486,95 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+  SmallVector<BasicBlock *, 8> ExitingBlocks;
+  TheLoop->getExitingBlocks(ExitingBlocks);
+
+  // Keep a record of all the exiting blocks with exact exit counts, as well as
+  // those with inexact counts.
```
As above, isn't this just all exiting blocks?

david-arm replied:
Done

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1442,6 +1486,95 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+    const SCEV *EC =
+        PSE.getSE()->getPredicatedExitCount(TheLoop, BB1, &Predicates);
+    if (isa<SCEVCouldNotCompute>(EC)) {
+      UncountableExitingBlocks.push_back(BB1);
```
Does it matter if BB1 has no out-of-loop successors?

david-arm replied:
If it's an exiting block at least one successor must be outside the loop. I've added an assert for this.

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1442,6 +1486,95 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+    return false;
+  }
+
+  // The only supported early exit loops so far are ones where the early
+  // exiting block is a unique predecessor of the latch block.
```
This answers my earlier question but that means the code in `canVectorizeMemory` is fragile.  Can an assert be added to `canVectorizeMemory` to ensure this is the case?

david-arm replied:
Done

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1442,6 +1486,95 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+    return false;
+  }
+
+  if (Reductions.size() || FixedOrderRecurrences.size()) {
+    reportVectorizationFailure(
+        "Found reductions or recurrences in early-exit loop",
+        "Cannot vectorize early exit loop with reductions or recurrences",
+        "RecurrencesInEarlyExitLoop", ORE, TheLoop);
+    return false;
+  }
```
This seems like a quick test that's worth having early before having to iterate across basic blocks etc...

david-arm replied:
Done

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1070,6 +1070,50 @@ bool LoopVectorizationLegality::canVectorizeMemory() {
     return false;
   }
 
+  // For loops with uncountable early exiting blocks that are not the latch
```
I don't think this is true, because you can have loops like this:

```
while.cond:                                       ; preds = %while.body, %entry
  %i.0 = phi i32 [ %start, %entry ], [ %inc, %while.body ]
  %inc = add i32 %i.0, 1
  %cmp.not = icmp eq i32 %inc, %end
  br i1 %cmp.not, label %cleanup, label %while.body

while.body:                                       ; preds = %while.cond
  %idxprom = zext i32 %inc to i64
  %arrayidx = getelementptr inbounds i32, ptr %p, i64 %idxprom
  %0 = load i32, ptr %arrayidx, align 4, !tbaa !6
  %cmp1 = icmp eq i32 %0, 7
  br i1 %cmp1, label %cleanup, label %while.cond, !llvm.loop !10

cleanup:                                          ; preds = %while.cond, %while.body
  %retval.0 = phi i32 [ 1, %while.body ], [ 0, %while.cond ]
  ret i32 %retval.0
```

We do see code like this being generated from C:

```
unsigned foo(int *p, unsigned start, unsigned end) {
  unsigned i = start;
  while (++i != end)
    if (p[i] == 7)
      return 1;
  return 0;
}
```

In this case the latch block either jumps to the beginning of the loop or takes an early exit.

david-arm committed:
Address review comments

* Reverted LoopAccessAnalysis changes in favour of new code in
LoopVectorizationLegality.cpp that checks if instructions are
safe to speculatively execute.
* Updated some comments and done minor code refactoring.
* Added new tests for demonstrating under which circumstances
we can speculatively execute calls and divides.
* Added new test where the early exit is in a conditional block.
* Added new test containing store instructions.
* Added new test containing loads after early exit.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1070,6 +1070,50 @@ bool LoopVectorizationLegality::canVectorizeMemory() {
+      }
+    }
+
+    // The vectoriser does not yet handle loops that may fault, but this will
+    // be improved in a follow-on patch.
```
Done

david-arm commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -446,6 +480,10 @@ class LoopVectorizationLegality {
   bool canVectorizeOuterLoop();
 
+  /// Returns true if this is a supported early exit loop that we can
+  /// vectorize.
```
Done

david-arm commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -551,6 +589,16 @@ class LoopVectorizationLegality {
+  /// uncountable exiting block that is not the latch.
+  bool HasSpeculativeEarlyExit = false;
+
+  /// Keeps track of all the exits with known or countable exit-not-taken
```
Adjusted the comment - hopefully makes more sense!

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1442,6 +1487,126 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+  SmallVector<const SCEVPredicate *, 4> Predicates;
+  for (BasicBlock *BB1 : ExitingBlocks) {
+    const SCEV *EC =
+        PSE.getSE()->getPredicatedExitCount(TheLoop, BB1, &Predicates);
```
Given `Predicates` is not used, can you pass `nullptr` here? From what I can see all the functions `getPredicatedExitCount` represent support not caring about the predicates.

david-arm replied:
I don't think that's true because if the loop really does require predicates then `getPredicatedExitCount` will return SCEVCouldNotCompute for the exit count and we will incorrectly treat this as an uncountable exiting block. If I pass in `nullptr` here the `simple_early_exit.ll` test file starts failing.

paulwalker-arm replied:
OK, I didn't see this parameter controlling any code paths but I do see an assert that would fail without it.  

What worried me was that you're clearing Predicates after iterating across all blocks and that made me assume it's contents didn't matter, but then I wondered why you need to manually clear it at all?

fhahn replied:
Could you add a comment here or below why predicates is passed and dropped below?

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1070,6 +1070,51 @@ bool LoopVectorizationLegality::canVectorizeMemory() {
+  // For loops with uncountable early exiting blocks that are not the latch
+  // it's necessary to perform extra checks, since the vectoriser is currently
+  // only capable of handling simple search loops.
+  if (IsEarlyExitLoop) {
```
Not sure what others think but I see `canVectorizeMemory` as being about load/store address analysis rather than "the only place to analyse load/store instructions". In this case you don't care about the addresses, you just want to guard against introducing side effects which I think can all be contained in `isVectorizableEarlyExitLoop`?

It kind of feels odd for `canVectorizeMemory` to depend on work done by `isVectorizableEarlyExitLoop`, when the "reason to reject an instruction" is largely the same.

fhahn replied:
IMO it would make sense to keep the checks together in `isVectorizableEarlyExitLoop` in particular with the extra checks recently added there

david-arm replied:
Done

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1442,6 +1487,126 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+
+  for (auto *BB : TheLoop->blocks())
+    for (auto &I : *BB)
+      if (!IsSafeOperation(&I)) {
```
Do early exiting loops introduce any special requirement here? I initially thought you'd restrict based on safe-to-speculate for only the blocks between the exit block and the latch block.  Which, given the current restriction, means just the latch block along the same lines as why you disallow loads in the latch block.

david-arm replied:
I think you have to restrict for all blocks up to and including the early exit too, because any operation after the vector lane that triggers the exit could be unsafe. For example, see test `@loop_contains_unsafe_div` where the early exit block looks like this:

```
loop:
  %index = phi i64 [ %index.next, %loop.inc ], [ 3, %entry ]
  %arrayidx = getelementptr inbounds i8, ptr %p1, i64 %index
  %ld1 = load i32, ptr %arrayidx, align 1
  %div = udiv i32 20000, %ld1
  %cmp = icmp eq i32 %div, 1
  br i1 %cmp, label %loop.inc, label %loop.end
```

It could be lane 1 of the comparison that triggers the early exit, but lane 3 could trap with divide-by-zero. If this was a normal loop where the divide was conditionally executed, i.e. something like

```
  for (int i = 0; i < n; i++) {
    if (mask[i]) {
      dst[i] = 20000 / src[i];
    }
  }
```

then the vectoriser can at least create a vector expression for the mask and select between `src` and a vector of ones to avoid unnecessarily creating an exception. However, in the test above the vectoriser has to perform the divide first in order to know what the mask should be.

I suppose in future improvements if we can prove that the `udiv` can be safely moved after the vector comparison (i.e. because the comparison does not depend upon the `udiv`) then we might be able to create a mask and select between src and 1.

For now, I thought it best to avoid over-complicating the initial early exit vectorisation work by bailing out on edge cases like this.

paulwalker-arm commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -551,6 +588,17 @@ class LoopVectorizationLegality {
+  /// uncountable exiting block that is not the latch.
+  bool HasSpeculativeEarlyExit = false;
+
+  /// Keep track of all the loop exiting blocks.
+  SmallVector<BasicBlock *, 4> CountableExitingBlocks;
+  SmallVector<BasicBlock *, 4> UncountableExitingBlocks;
+
+  /// Keep track of the destinations of all uncountable exits.
+  SmallVector<BasicBlock *, 4> UncountableExitBlocks;
```
Of the three only `UncountableExitingBlocks` looks to be used by this patch.  I'd just like confirmation this it intentional and will be built upon in a following PR.

david-arm replied:
Yeah, these will be used in follow-on patches. However, if you prefer to keep this initial patch clean I can remove them and add them later. I was just trying to reduce the size of follow-on patches by moving it into this patch that's all.

fhahn commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -377,6 +377,24 @@ class LoopVectorizationLegality {
+
+  /// Returns the destination of a speculative early exiting block.
+  BasicBlock *getSpeculativeEarlyExitBlock() const {
+    assert(getUncountableExitBlocks().size() == 1 &&
```
Both `getSpeculativeEarlyExitBloc` and `getUncountableExitingBlocks` assert that there's exactly one entry. Would it be simpler to store the pointers directly, rather than in a 1 element vector?

david-arm replied:
I could do that, and would be happy to do so if people prefer. I wrote it this way because in future I thought it's possible we may support multiple early exits. @paulwalker-arm @huntergr-arm any preference?

fhahn commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -436,7 +470,7 @@ class LoopVectorizationLegality {
   /// legal to vectorize the code, considering only memory constrains.
   /// Returns true if the loop is vectorizable
-  bool canVectorizeMemory();
+  bool canVectorizeMemory(bool IsEarlyExitLoop);
```
Document argument

david-arm replied:
Done

fhahn commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -446,6 +480,9 @@ class LoopVectorizationLegality {
   /// specific checks for outer loop vectorization.
   bool canVectorizeOuterLoop();
 
+  /// Returns true if this is an early exit loop that can be vectorized.
```
It would probably be helpful to summarize the required conditions here (or somewhere else)

david-arm replied:
Done

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1442,6 +1487,126 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+  // Check all instructions in the loop to see if they could potentially
+  // generate exceptions or have side-effects.
+  auto IsSafeOperation = [](Instruction *I) -> bool {
+    // Is this a divide?
```
The comment doesn't seem to match the code below?

david-arm replied:
Hopefully this reads better now!

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1070,6 +1070,51 @@ bool LoopVectorizationLegality::canVectorizeMemory() {
+               getUncountableExitingBlocks()[0] &&
+           "Expected latch predecessor to be the early exiting block");
+
+    for (Instruction &I : *LatchBB) {
```
Do we have any restrictions that there is only the latch after the early exit? If not, all blocks between early exit and latch would need to be check?

david-arm replied:
Yeah, if you look in `isVectorizableEarlyExitLoop` you'll see this code:

```
  // The only supported early exit loops so far are ones where the early
  // exiting block is a unique predecessor of the latch block.
  BasicBlock *LatchPredBB = LatchBB->getUniquePredecessor();
  if (LatchPredBB != getUncountableExitingBlocks()[0]) {
    reportVectorizationFailure("Early exit is not the latch predecessor",
                               "Cannot vectorize early exit loop",
                               "EarlyExitNotLatchPredecessor", ORE, TheLoop);
    return false;
  }
```

which is why a few lines above I am able to add this assert:

```
    assert(LatchBB->getUniquePredecessor() ==
               getUncountableExitingBlocks()[0] &&
           "Expected latch predecessor to be the early exiting block");
```

Although if I move the new code from `canVectorizeMemory` to `isVectorizableEarlyExitLoop` I can get rid of the assert.

I added this restriction as a simple optimisation to avoid messier CFGs, such as conditionally executed blocks after the early exit:

```
loop:
...
  br i1 %cmp, label %early.exit, label %loop.bb1

loop.bb1:
...
  br i1 %cmp2, label %do.stuff, label %loop.latch

do.stuff:
...
  br label %loop.latch

loop.latch:
...
  br i1 %cmp3, label %normal.exit, label %loop
```

In practice we can't yet permit any useful work after the early exit anyway until we can teach the vectoriser that subsequent blocks are effectively if-converted and require a mask, and the operations would need to occur before the early exit.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1442,6 +1487,126 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
 }
 
+bool LoopVectorizationLegality::isVectorizableEarlyExitLoop() {
+  // At least one of the exiting blocks must be the latch.
```
Below just checks if there's a latch, not if it is exiting?

david-arm replied:
Hmm, I should move the comment lower down to the code where it's checked:

```
  if (isa<SCEVCouldNotCompute>(
          PSE.getSE()->getPredicatedExitCount(TheLoop, LatchBB, &Predicates))) {
    reportVectorizationFailure(
        "Cannot determine exact exit count for latch block",
        "Cannot vectorize early exit loop",
        "UnknownLatchExitCountEarlyExitLoop", ORE, TheLoop);
    return false;
  }
```

There is a test for this - see `@early_exit_infinite_loop`

david-arm replied:
Done

david-arm committed:
Address review comments

* Combined changes from canVectorizeMemory into
isVectorizableEarlyExitLoop.
* Documented restrictions for isVectorizableEarlyExitLoop in
header file.
* Added reportVectorizationFailure call in LoopVectorize when
we bail out.
* General code refactor, clean-up.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9802,6 +9802,12 @@ bool LoopVectorizePass::processLoop(Loop *L) {
 
+  if (LVL.hasSpeculativeEarlyExit()) {
+    LLVM_DEBUG(dbgs() << "LV: Not vectorizing: Auto-vectorization of early "
+                      << "exit loops is not yet supported.\n");
```
Done

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1442,6 +1487,126 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+  // The only supported early exit loops so far are ones where the early
+  // exiting block is a unique predecessor of the latch block.
+  BasicBlock *LatchPredBB = LatchBB->getUniquePredecessor();
+  if (!LatchPredBB || LatchPredBB != getUncountableExitingBlocks()[0]) {
```
Done

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1070,6 +1070,51 @@ bool LoopVectorizationLegality::canVectorizeMemory() {
+    // The vectoriser cannot handle loads that occur after the early exit block.
+    BasicBlock *LatchBB = TheLoop->getLoopLatch();
+    assert(LatchBB->getUniquePredecessor() ==
+               getUncountableExitingBlocks()[0] &&
```
Done

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1442,6 +1442,157 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+  // The vectoriser cannot handle loads that occur after the early exit block.
+  assert(LatchBB->getUniquePredecessor() == getSpeculativeEarlyExitingBlock() &&
+         "Expected latch predecessor to be the early exiting block");
+  for (Instruction &I : *LatchBB) {
+    if (I.mayReadFromMemory()) {
+      reportVectorizationFailure(
+          "Loads not permitted after early exit",
+          "Cannot vectorize early exit loop with loads after early exit",
+          "LoadsAfterEarlyExit", ORE, TheLoop);
+      return false;
+    }
+    // Any other problematic instructions should have been caught earlier.
+    assert(!I.mayWriteToMemory() && !I.mayThrow() && !I.mayHaveSideEffects() &&
+           "Unexpected instructions in latch block of early exit loop");
+  }
```
I think the `!isDereferenceableReadOnlyLoop` check below makes this loop redundant and thus can be removed.

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1442,6 +1442,157 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+      }
+    }
+
+  // At least one of the exiting blocks must be the latch.
```
This comment does not seem to match the code.

paulwalker-arm reviewed (approved):
A few final comments but otherwise this looks good to me.

david-arm commented:
Rebased to fix the conflict (that wasn't a conflict) warning

david-arm committed:
Address review comments

* Remove code to check loads after early exit, since at the
moment this is redundant. We only vectorise if we can prove
that none of the loads will fault, so there is no harm in
speculatively executing. Any outside users of a load in the
latch block must have come from a normal exit and so we
will extract the last lane of the vectorised in the same way
as a normal loop.
* Re-wrote the test loop_contains_load_after_early_exit to
show it is legal to vectorise when the load in the latch
block is dereferenceable. The test also contains an outside
use of that load. At least once a patch lands to start
vectorising this loop it will be obvious whether the code
is correct or not.
* Updated a comment.

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/simple_early_exit.ll:
```diff
@@ -0,0 +1,1969 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt -S < %s -p loop-vectorize -mattr=+sve -debug-only=loop-vectorize \
```
There's nothing target specific in this test at the moment, can it be moved out so we have test coverage for the legality checks independent of the target?

david-arm replied:
Sure, seems reasonable. We will need a few SVE specific tests once we start vectorising these loops just to ensure the cost model decides it's worth vectorising, but that's fine!

david-arm committed:
Address review comments

* Move simple_early_exit.ll test file to top level directory and
changed command line to be target-independent. As a result of this
I had to update the tests due to some alignment differences.

fhahn commented on llvm/test/Transforms/LoopVectorize/simple_early_exit.ll:
```diff
@@ -0,0 +1,1941 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
```
This will need REQUIRES:asserts I think

david-arm replied:
Ah yes, you're right! Good spot. I'll fix asap.

fhahn commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -551,6 +602,17 @@ class LoopVectorizationLegality {
+
+  /// Indicates whether this loop has a speculative early exit, i.e. an
+  /// uncountable exiting block that is not the latch.
+  bool HasSpeculativeEarlyExit = false;
```
For consistency, with SCEV terminology and the variables below, `Speculative` -> `Uncountable`?

fhahn commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -551,6 +602,17 @@ class LoopVectorizationLegality {
+  /// uncountable exiting block that is not the latch.
+  bool HasSpeculativeEarlyExit = false;
+
+  /// Keep track of all the loop exiting blocks.
```
Is this comment accurate? AFAICT they only keep track of a single countable and uncountable exiting block, *if* the backedge taken count is not computable.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1445,6 +1445,145 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+    }
+
+  // The latch block must have a countable exit.
+  if (isa<SCEVCouldNotCompute>(
```
Move simpler checks earlier before iterating over the whole loop? Can also check if LatchBB is in the collected countable exiting blocks?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1445,6 +1445,145 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+
+  LLVM_DEBUG(
+      dbgs()
+      << "LV: Found an early exit. Retrying with speculative exit count.\n");
```
Would be good to be consistent with SCEV terminology, i.e. symbolic max BTC. Or more specifically the computable exiting count for the loop latch, as with the current restrictions `getSymbolicMaxBackedgeTakenCount() == >getPredicatedExitCount(TheLoop, Latch)`?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1445,6 +1445,145 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+        return false;
+      }
+
+      BasicBlock *BB2;
```
clearer to name ExitBlock?

david-arm commented:
Hi @fhahn, I'll take a look at addressing your comments in a follow-up patch. I had used the term speculative for things like `HasSpeculativeEarlyExit` to try and indicate that even the loop has to have more than just an uncountable early exit, because the exit count from the latch needs to be countable too in order for this flag to be true. But perhaps that's unnecessary and a bit pedantic and I think I'm happy to rename for consistency.
